observer/baseNodeProvider.go

```diff
@@ -28,26 +30,214 @@ func (bop *baseNodeProvider) initNodesMaps(nodes []*data.NodeData) error {
 		newNodes[shardId] = append(newNodes[shardId], observer)
 	}
 
-	bop.mutNodes.Lock()
-	bop.nodes = newNodes
-	bop.allNodes = initAllNodesSlice(newNodes)
-	bop.mutNodes.Unlock()
+	bnp.mutNodes.Lock()
+	bnp.nodesMap = newNodes
+	bnp.syncedNodes = initAllNodesSlice(newNodes)
+	bnp.outOfSyncNodes = make([]*data.NodeData, 0)
+	bnp.mutNodes.Unlock()
 
 	return nil
 }
 
+// GetAllNodesWithSyncState will return the merged list of active observers and out of sync observers
+func (bnp *baseNodeProvider) GetAllNodesWithSyncState() []*data.NodeData {
+	bnp.mutNodes.RLock()
+	defer bnp.mutNodes.RUnlock()
+
+	nodesSlice := make([]*data.NodeData, 0)
+	for _, node := range bnp.syncedNodes {
+		nodesSlice = append(nodesSlice, node)
+	}
+	for _, node := range bnp.outOfSyncNodes {
+		nodesSlice = append(nodesSlice, node)
+	}
+
+	return nodesSlice
+}
+
+// UpdateNodesBasedOnSyncState will handle the nodes lists, by removing out of sync observers or by adding back observers
+// that were previously removed because they were out of sync.
+func (bnp *baseNodeProvider) UpdateNodesBasedOnSyncState(nodesWithSyncStatus []*data.NodeData) {
+	bnp.mutNodes.Lock()
+	defer bnp.mutNodes.Unlock()
+
+	shardIDs := getSortedSliceIDsSlice(bnp.nodesMap)
+	syncedNodes, outOfSyncNodes, err := computeSyncedAndOutOfSyncNodes(nodesWithSyncStatus, shardIDs)
+	if err != nil {
+		log.Error("cannot update nodes based on sync state", "error", err)
+		return
+	}
+
+	syncedNodesMap := nodesSliceToShardedMap(syncedNodes)
+
+	if len(bnp.syncedNodes) == len(syncedNodes) && len(outOfSyncNodes) == 0 {
```

Comment:
perhaps this check can be moved before L71 to avoid that computation, as it would be useless if we return from here

Reply:
right. refactored